### PR TITLE
#2640 Improve efficiency of profile loading code

### PR DESCRIPTION
**Pull request by yashdev9274**
Used the 'async' and 'await' keywords, as well as the 'fs.promises' API, can improve the efficiency of the code by allowing it to perform multiple operations concurrently, rather than waiting for each operation to finish before starting the next one. This can result in faster execution times and reduced server load.

## Fixes Issue

Fixed issue #2640

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

image

## Note to reviewers

Additionally, I have used pagination or filtering to limit the number of documents retrieved from the Profile collection can also improve efficiency by reducing the amount of data that needs to be processed.

Finally, wrapping the JSON.parse function in a try-catch block can help handle any errors that might occur during parsing, which can help improve the reliability and robustness of the code.

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/2641"></a>

### CODE REVIEW

File: pages/api/users/index.js [hunk 1]
```diff
@@ -6,23 +6,28 @@ import Profile from "../../../models/Profile";
...
-    .filter((item) => item.includes("json"));
+  const files = await fs.promises.readdir(directoryPath);
+
+  const users = await Promise.all(
+    files
+      .filter((item) => item.includes("json"))
+      .map(async (file) => {
+        const filePath = path.join(directoryPath, file);
+        try {
+          const data = await fs.promises.readFile(filePath, "utf8");
```
Comment (eddiejaoude) [inline pages/api/users/index.js:19]: I think this will block the `map`?

File: pages/api/users/index.js [hunk 2]
```diff
@@ -6,23 +6,28 @@ import Profile from "../../../models/Profile";
...
 export default async function handler(req, res) {
   await connectMongo();
+
   const directoryPath = path.join(process.cwd(), "data");
-  const files = fs
-    .readdirSync(directoryPath)
-    .filter((item) => item.includes("json"));
+  const files = await fs.promises.readdir(directoryPath);
+
+  const users = await Promise.all(
```
Comment (eddiejaoude) [inline pages/api/users/index.js:13]: The `Promise.all` should have a try/catch as it could fail

**Review by eddiejaoude — CHANGES_REQUESTED** [review]
Great idea Yash 👍 

I left some inline comments.

File: pages/api/users/index.js [hunk 3]
```diff
@@ -6,23 +6,28 @@ import Profile from "../../../models/Profile";
...
+        const filePath = path.join(directoryPath, file);
+        try {
+          const data = await fs.promises.readFile(filePath, "utf8");
+          return {
+            ...JSON.parse(data),
+            username: file.split(".")[0],
+          };
+        } catch (e) {
+          console.log(`ERROR loading profile "${filePath}"`);
+          return null;
```
Comment (eddiejaoude) [inline pages/api/users/index.js:26]: `flatMap` was used to remove any files that did not load, but now there will be a profile that has `null` in it?

File: pages/api/users/index.js (hunk 1 above)
Comment (yashdev9274) [reply pages/api/users/index.js:19]: Then what changes should I make in here?

Comment (eddiejaoude) [reply pages/api/users/index.js:19]: The async reading of the files should happen async, then loaded into the promise all

**Comment by swiftugandan** [comment]
Perhaps you should consider looking at other `fs.*` calls and update them to use `fs.promises`, not just the user api end-point

**Comment by eddiejaoude** [comment]
@yashdev9274 thank you for your contribution and efforts, but I am going to close this pull request as I think it needs more thought around how we solve this issue

**Comment by yashdev9274** [comment]
@eddiejaoude thank you for your feedback, I'll be available for the discussion on respective issue , do let me know you views

**Pull request closed without merging**